### ISSUE TITLE
Allow (temporary) changing of managed settings

### ISSUE THREAD

**Issue by portiatrivisonno** [foxyproxy/browser-extension#132]
I'd like to be able to change the setting provided by the managed storage (policies/manifest) during runtime.

I'm often deploying a fresh install of Firefox with FoxyProxy for proxy management.  I use policies to deploy the initial configuration but would like to allow the user (including myself) to change the settings (more precisely the active proxy) afterwards.  This, however, seems currently not possible.

There was already a similar request in [1] using Chrome's "recommended policy".  But I'm not aware of similar mechanism in Firefox.  I could think of a setting "allow overwriting managed config", which, if enabled, allows changing the settings even in a "managed environment".  (A quick glance at the code showed a lot of `if (pref.managed) return;` patterns.)

Personally, I don't care much whether the user's changes are permanent or only for the session; I guess the latter is probably preferred by most people.

Thanks in advance!

[1] https://github.com/foxyproxy/browser-extension/issues/95#issuecomment-1859139694

**Comment by erosman** [comment]
As mentioned in https://github.com/foxyproxy/browser-extension/issues/95#issuecomment-1859171736

> That would conflict with https://github.com/foxyproxy/browser-extension/issues/42#issuecomment-1830407101
> 
> The managed Enterprise Policy is set to give control to the admin and thus remove the control from the user.

It appears that the request  is for a feature to apply a "default settings", rather than "managed storage/enterprise policy".

I have to look into the possibilities, if there is a popular demand for it.

**Comment by portiatrivisonno** [comment]
> As mentioned in [#95 (comment)](https://github.com/foxyproxy/browser-extension/issues/95#issuecomment-1859171736)
> 
> > That would conflict with [#42 (comment)](https://github.com/foxyproxy/browser-extension/issues/42#issuecomment-1830407101)
> > The managed Enterprise Policy is set to give control to the admin and thus remove the control from the user.

"conflict" might be a misleading word here.  Given an option "allow overwriting managed config" (which is set in the policy), the admin can choose to give the user some control back or deny them all control.

> It appears that the request is for a feature to apply a "default settings", rather than "managed storage/enterprise policy".
> 
> I have to look into the possibilities, if there is a popular demand for it.

For what its worth, uBlockOrigin offers something similar [1,2]. They allow the user to change settings during runtime but reset them on every launch. Maybe it's worth a look how they implemented that.

[1] https://github.com/gorhill/uBlock/wiki/Deploying-uBlock-Origin
[2] https://github.com/gorhill/uBlock/wiki/Deploying-uBlock-Origin:-configuration

**Comment by erosman** [comment]
The feature can be added if there is a popular demand for it.

**Comment by voronind-com** [comment]
> The feature can be added if there is a popular demand for it.

Yes, please! <3

**Comment by erosman** [comment]
I have been thinking about this request.

- It is possible to add a property to the `policies.json` e.g. `editable: true`
- In that case, FoxyProxy loads the Enterprise policy and if `editable: true`, it would allow changing of the options

However, all changes would be for the session only since the next time browser starts, it will reset everything to the values in  `policies.json`.

It should not be a major issue for most, but in MV3 background script is unloaded and every time it is restarted, it would change everything back to the  `policies.json`  values.

Chrome FoxyProxy is MV3. Firefox is still MV2 but will be MV3 in future.

**Comment by voronind-com** [comment]
Sounds nice. But maybe inverse it so it's `locked = true`? Just a suggestion.

**Comment by erosman** [comment]
> Sounds nice. But maybe inverse it so it's `locked: true`? Just a suggestion.

That would make the  default behavior editable, which goes against the default concept of the enterprise policy.


> Also FF will keep supporting MV2.

To start with, the feature will have the following consequence for the Chrome (i.e. the majority of) users.
Firefox has already started to support MV3 and maybe within a year all will change to MV3.

### The problem for users on MV3

- start with enterprise policy values
- make changes
- go make a cup of tea
- come back to see the settings are back to the  enterprise policy values

**Comment by erosman** [comment]
Enterprise policy is not meant for editing. Therefore, the feature request is not suitable for enterprise policy.

Please open a new issue if alternative method is required.

![image](https://github.com/user-attachments/assets/9626fc46-1757-4bdc-a2a2-54ccb682bf1b)

**Comment by erosman** [comment]
Based on the concept, enterprise policy is not meant for user editing, thus the feature request is not suitable for an enterprise policy.

Please open a new issue if alternative method is required.

![image](https://github.com/user-attachments/assets/9626fc46-1757-4bdc-a2a2-54ccb682bf1b)

**Comment by voronind-com** [comment]
Dude if you ask pointless AI please at least give a second of a human though. Users can't edit the policy itself (the json file), not in-app configuration.

**Comment by erosman** [comment]
Please show a reference to where Chrome and Firefox state that Enterprise Policies are editable by the users.

**Comment by voronind-com** [comment]
> Please show a reference to where Chrome and Firefox state that Enterprise Policies are editable by the users.

I'll only give FF ref but Chromium should be the same. This is an example: https://mozilla.github.io/policy-templates/#dnsoverhttps

There's a 'locked' key, which is why I asked you to inverse the logic as by default values can be changed. There are many others, just search by word 'locked'.

Also there are those preferences: https://support.mozilla.org/en-US/kb/customizing-firefox-using-autoconfig#w_functions-of-autoconfig

See the "lockPref" and how it differs to others.

Sorry for being rude, I was just too saddened by the "closed by ai" response.

**Comment by erosman** [comment]
[3rdparty enterprise policy](https://mozilla.github.io/policy-templates/#3rdparty) shows the option for extensions.

Enterprise polices are accessed via `storage.managed` API.

### [storage.managed](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage/managed)

> A [storage.StorageArea](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage/StorageArea) object that represents the `managed` storage area. Items in `managed` storage are set by the domain administrator or other native applications installed on the user's computer and are read-only for the extension. Trying to modify this storage area results in an error.

**Comment by voronind-com** [comment]
> [3rdparty enterprise policy](https://mozilla.github.io/policy-templates/#3rdparty) shows the option for extensions.
> 
> Enterprise polices are accessed via `storage.managed` API.
> ### [storage.managed](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage/managed)
> 
> > A [storage.StorageArea](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage/StorageArea) object that represents the `managed` storage area. Items in `managed` storage are set by the domain administrator or other native applications installed on the user's computer and are read-only for the extension. Trying to modify this storage area results in an error.

Sure, this is like where to install your app. Sure thing it is not controlled by the app itself. All I'm saying is that this area is flexible and there are examples of both casrs.

**Comment by erosman** [comment]
As mentioned, enterprise policy is not suitable for this feature.

> Please open a new issue if an alternative method is required.

**Comment by linsui** [comment]
I want to set tab proxy when managed storage is enabled. This is the behavior of uBO. All the changes are temporary.

**Comment by voronind-com** [comment]
> I want to set tab proxy when managed storage is enabled. This is the behavior of uBO. All the changes are temporary.

The developer "proofed" that this is wrong by attaching a prompt result from Google text generator.
Tho I highly agree that uBO should be a standard in terms of policy configurations.

**Comment by erosman** [comment]
> This is the behavior of uBO. All the changes are temporary.

Is it documented?

**Comment by erosman** [comment]
I had a quick look through the uBO docs and code. There doesn’t seem to be anything about editable enterprise policy.
It also appear that the editable aspect might be a consequence of the implementation method.
I didn’t find anything specific e.g. a flag that allows user editing.

See also:
- [managed_storage.json](https://github.com/gorhill/uBlock/blob/efd1d7d171eab505b22a7036fa51a4fcc4e7ad2f/platform/common/managed_storage.json)

**Comment by linsui** [comment]
Yep, I thought it's just a implementation detail. But I thought it's useful. The autoConfig in Firefox has similiar logic that an value can be set as default or locked.

**Comment by erosman** [comment]
> I want to set tab proxy when managed storage is enabled. This is the behavior of uBO. All the changes are temporary.

@linsui As Tab Proxy data is not saved in `storage.managed` (sessions only), if you post a request, I will change that to allow tab proxy.

**Comment by linsui** [comment]
https://github.com/foxyproxy/browser-extension/issues/172 Thanks a lot!
